### ISSUE TITLE
GA seems to "accumulate" data as well

### ISSUE THREAD

**Issue by DBAFC** [run-ai/runai#25]
Hi

I'm trying to train an image segmentation model (U-net) in keras using your gradient accumulation method. It seems to work (model compiles and training starts) but as training progresses it gets slower. The problem is (seemingly) that the data is "accumulated" or "copied" at the start of each epoch, such that the training data set becomes larger for each epoch (see example below).

Epoch 1/400
768/768 [==============================] - 1012s 1s/step - loss: 0.6437 - mean_IoU: 0.5662 - accuracy: 0.7541 - val_loss: 0.1128 - val_mean_IoU: 0.6391 - val_accuracy: 0.9152

Epoch 00001: val_loss improved from inf to 0.11280, saving model to unet_1__lr_10E-04_drop_02_reg_10E-05.hdf5
Epoch 2/400
2208/2208 [==============================] - 1389s 629ms/step - loss: 0.3432 - mean_IoU: 0.6955 - accuracy: 0.9326 - val_loss: 0.2322 - val_mean_IoU: 0.7350 - val_accuracy: 0.9470

Epoch 00002: val_loss did not improve from 0.11280
Epoch 3/400
4188/4188 [==============================] - 2603s 622ms/step - loss: 0.2730 - mean_IoU: 0.7266 - accuracy: 0.9330 - val_loss: 0.5292 - val_mean_IoU: 0.7454 - val_accuracy: 0.9557

Epoch 00003: val_loss did not improve from 0.11280
Epoch 4/400
6548/6548 [==============================] - 4043s 617ms/step - loss: 0.1635 - mean_IoU: 0.7716 - accuracy: 0.9572 - val_loss: 0.1800 - val_mean_IoU: 0.7883 - val_accuracy: 0.9618
 
I initialize the model like so:

import runai.ga.keras as rga

model = unet((input_shape), droprate = droprate, num_classes=5)
optimizer = rga.optimizers.Optimizer(Adam(lr),steps=5)
model.compile(optimizer = optimizer,loss = loss, metrics = ['mean_IoU','accuracy'])

and training is done using the model.fit() function:

history = model.fit(train_gen,validation_data=val_gen, epochs=400, callbacks = callbacks, verbose = True)

I use a custom datagenerator which works fine when I don't use GA and the problem persists whether I use model.fit() or model.fit_generator()

 Is this expected behavior when using a data-generator or do you have any idea as to what is happening. 

Running your VGG test script does not produce the same issue.

Tensorflow version = 1.14.0

Keras version =  2.3.1

**Comment by levosos** [comment]
Hi @DBAFC, thanks for submitting this detailed issue!
You are correct, this is not the desired behavior, and it indeed seems like data is being "accumulated" over epochs for some reason.

I think it's related to the data generator.
What did you use for `batch_size` and `steps_per_epoch` when created the data generator?
We have tested it using a data generator and it should work. Please take a look at [this example](https://github.com/run-ai/runai/blob/master/examples/elastic/keras/cifar10.py#L92) (it's an example for elasticity but it should be very similar to GA).

At first look, it seems like you are doing things fine.
Therefore, it'd be great if you shared your code here entirely, or at least a minimal version so that we could reproduce the issue and help in fixing that.

Thanks!

**Comment by DBAFC** [comment]
Sorry for the long response time. 

So I think i figured it out. The problem was indeed related to the generator. I'm working with quite large images with a lot of background, so the way I read the data is sampling patches from areas containing relevant information. Indexes of the areas are stored in lists which are assigned to a variable of indexes, and during training I am adding random background patches to the list. For some reason, when I use GA the list sizes increase because (seemingly) background patches are added to the original list (as opposed to the variable. Thereby changing it's length). For some reason this did not happen when not using GA. So i solved the problem by making a copy of the list and assigning that the variable, instead of assigning it directly.

Thanks for your help.

**Comment by levosos** [comment]
I'm glad you figured it out. 
Let us know if we can help with anything else in the future!
